v23/context/context.go

```diff
@@ -303,8 +303,12 @@ func WithRootCancel(parent *T) (*T, CancelFunc) {
 		// Forward the cancelation from the root context to the newly
 		// created context.
 		go func() {
-			<-rootCtx.Done()
-			cancel()
+			select {
+			case <-rootCtx.Done():
+				cancel()
+			case <-ctx.Done():
+				cancel()
```

Comment:
I believe this line isn't really needed since it's a noop. I left this here to be consistent though.

Reply:
I think you need it for the case that the child context is canceled but the root is not - this will be the common case for servers that forward requests.

Reply:
Ah yes I mean the `cancel()` on line 310. The `case <-ctx.Done():` for sure is needed but I believe if the child context is canceled or finished, it'll enter the select case and run cancel(), which I believe would just be a noop.